File: python/src/rate_limit.py
import time
from threading import Lock
from typing import Dict, Tuple

from starlette.requests import Request

WINDOW_SECONDS = 60.0


def client_ip(request: Request) -> str:
    forwarded = request.headers.get("x-forwarded-for")
    if forwarded:
        return forwarded.split(",")[0].strip()
    if request.client:
        return request.client.host
    return "unknown"


class FixedWindowRateLimiter:
    """Per-client fixed-window request counter (thread-safe)."""

    def __init__(self, limit_per_minute: int) -> None:
        self.limit_per_minute = limit_per_minute
        self._windows: Dict[str, Tuple[float, int]] = {}
        self._lock = Lock()

    def allow(self, client: str) -> bool:
        if self.limit_per_minute == 0:
            return True

        now = time.monotonic()
        with self._lock:
            window_start, count = self._windows.get(client, (now, 0))
            if now - window_start >= WINDOW_SECONDS:
                window_start = now
                count = 0
            if count >= self.limit_per_minute:
                self._windows[client] = (window_start, count)
                return False
            self._windows[client] = (window_start, count + 1)
            return True

File: rust/src/rate_limit.rs
use std::{
    collections::HashMap,
    sync::Mutex,
    time::{Duration, Instant},
};

const WINDOW: Duration = Duration::from_secs(60);

pub struct RateLimiter {
    limit_per_minute: u32,
    windows: Mutex<HashMap<String, (Instant, u32)>>,
}

impl RateLimiter {
    pub fn new(limit_per_minute: u32) -> Self {
        RateLimiter {
            limit_per_minute,
            windows: Mutex::new(HashMap::new()),
        }
    }

    pub fn allow(&self, client: &str) -> bool {
        if self.limit_per_minute == 0 {
            return true;
        }

        let mut guard = match self.windows.lock() {
            Ok(guard) => guard,
            Err(err) => {
                log::error!("Rate limiter lock poisoned: {err}");
                return false;
            }
        };

        let now = Instant::now();
        let entry = guard.entry(client.to_string()).or_insert((now, 0));
        if now.duration_since(entry.0) >= WINDOW {
            *entry = (now, 0);
        }
        if entry.1 >= self.limit_per_minute {
            return false;
        }
        entry.1 += 1;
        true
    }
}

pub fn client_ip(req: &actix_web::HttpRequest) -> String {
    if let Some(forwarded) = req
        .headers()
        .get("X-Forwarded-For")
        .and_then(|value| value.to_str().ok())
    {
        return forwarded
            .split(',')
            .next()
            .unwrap_or(forwarded)
            .trim()
            .to_string();
    }
    req.connection_info()
        .peer_addr()
        .unwrap_or("unknown")
        .to_string()
}

#[cfg(test)]
mod tests {
    use super::RateLimiter;

    #[test]
    fn disabled_limiter_allows_all() {
        let limiter = RateLimiter::new(0);
        assert!(limiter.allow("127.0.0.1"));
        assert!(limiter.allow("127.0.0.1"));
    }

    #[test]
    fn enforces_limit_per_window() {
        let limiter = RateLimiter::new(2);
        assert!(limiter.allow("client-a"));
        assert!(limiter.allow("client-a"));
        assert!(!limiter.allow("client-a"));
        assert!(limiter.allow("client-b"));
    }
}
